test/extensions/filters/http/cache/cache_filter_test.cc

```diff
@@ -762,6 +762,104 @@ TEST_F(CacheFilterTest, FilterDeletedBeforePostedCallbackExecuted) {
   }
 }
 
+TEST_F(CacheFilterTest, LocalReplyDuringLookup) {
+  request_headers_.setHost("LocalReplyDuringLookup");
+  {
+    // Create filter for request 1.
+    CacheFilterSharedPtr filter = makeFilter(simple_cache_);
+
+    testDecodeRequestMiss(filter);
+
+    // Encode response headers.
+    EXPECT_EQ(filter->encodeHeaders(response_headers_, true), Http::FilterHeadersStatus::Continue);
+
+    filter->onStreamComplete();
+    EXPECT_THAT(lookupStatus(), IsOkAndHolds(LookupStatus::CacheMiss));
+
+    filter->onDestroy();
+  }
+  {
+    // Create filter for request 2.
+    CacheFilterSharedPtr filter = makeFilter(simple_cache_);
+
+    // Call decode headers to start the cache lookup, which should immediately post the callback to
+    // the dispatcher.
+    EXPECT_EQ(filter->decodeHeaders(request_headers_, true),
+              Http::FilterHeadersStatus::StopAllIterationAndWatermark);
+
+    // While the lookup callback is still on the dispatcher, simulate a local reply.
+    Envoy::Http::TestResponseHeaderMapImpl local_response_headers{{":status", "503"}};
+    EXPECT_EQ(filter->encodeHeaders(local_response_headers, true),
+              Envoy::Http::FilterHeadersStatus::Continue);
+
+    // Make sure that the filter doesn't try to encode the cached response after processing the
+    // local reply.
+    EXPECT_CALL(decoder_callbacks_, continueDecoding).Times(0);
+    EXPECT_CALL(decoder_callbacks_, encodeHeaders_).Times(0);
+
+    // Run events on the dispatcher so that the lookup callback is invoked after the local reply.
+    dispatcher_->run(Event::Dispatcher::RunType::Block);
+
+    filter->onStreamComplete();
+    EXPECT_THAT(lookupStatus(), IsOkAndHolds(LookupStatus::RequestIncomplete));
+    filter->onDestroy();
+    filter.reset();
+  }
+}
+
+using CacheFilterDeathTest = CacheFilterTest;
```

Comment:
what does this nickname do?

Reply:
Tells googletest it's a death test for threading purposes. See https://google.github.io/googletest/advanced.html#death-test-naming

Reply:
Done.